Intermittent test failure of compression test

The compression test using CLI can sometimes fail because it checks for the compressed file before the gzip process has completed. The callback is currently called when the backup child process exits, but should not be called until the gzip process has exited.

- couchbackup/couchrestore version(s) that are affected by this issue.
> 2.0.0-SNAPSHOT test issue
- Node.js version.
> Any
- A sample of the code or command line statement that demonstrates the issue.

Running `should backup animaldb to a compressed file.Compression tests using CLI should backup animaldb to a compressed file (from Compression tests using CLI)`

>AssertionError: The backup file should be gz compressed.
    at Object.assertGzipFile (citestutils.js:381:12)
    at test.js:87:15
    at ChildProcess.<anonymous> (citestutils.js:182:9)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:215:12)

--- a/citest/citestutils.js
+++ b/citest/citestutils.js
@@ -59,6 +59,7 @@ function params(params, o) {
 }
 
 function testBackup(params, databaseName, outputStream, callback) {
+  var gzip;
   var backupStream = outputStream;
 
   // Pipe via compression if requested
@@ -70,10 +71,14 @@ function testBackup(params, databaseName, outputStream, callback) {
       backupStream.pipe(outputStream);
     } else {
       // Spawn process for gzip
-      const gzip = spawn('gzip', [], {'stdio': ['pipe', 'pipe', 'inherit']});
+      gzip = spawn('gzip', [], {'stdio': ['pipe', 'pipe', 'inherit']});
       // Pipe the streams as needed
       gzip.stdout.pipe(outputStream);
       backupStream = gzip.stdin;
+      // register an error handler
+      gzip.on('error', function(err) {
+        callback(err);
+      });
     }
   }
 
@@ -101,7 +106,6 @@ function testBackup(params, databaseName, outputStream, callback) {
     backup.on('exit', function(code) {
       try {
         assert.equal(code, 0, 'The backup should exit normally.');
-        callback();
       } catch (err) {
         callback(err);
       }
@@ -109,6 +113,16 @@ function testBackup(params, databaseName, outputStream, callback) {
     backup.on('error', function(err) {
       callback(err);
     });
+    // Call done when the last child process exits - could be gzip or backup
+    if (gzip) {
+      gzip.on('exit', function(code) {
+        callback();
+      });
+    } else {
+      backup.on('exit', function(code) {
+        callback();
+      });
+    }
   }
 }
 

